walk/walk.go

```diff
@@ -217,6 +217,54 @@ func Walk2(c *config.Config, cexts []config.Configurer, dirs []string, mode Mode
 		return errors.Join(w.errs...)
 	}
 
+	// Visit additional directories that extensions requested for indexing.
+	// Don't visit subdirectories recursively, even when recursion is enabled.
+	w.mode = UpdateDirsMode
+	for len(w.relsToVisit) > 0 {
+		// Don't simply range over relsToVisit. We may append more.
+		relToVisit := w.relsToVisit[0]
+		w.relsToVisit = w.relsToVisit[1:]
+
+		// Make sure to visit prefixes of relToVisit as well so we apply
+		// configuration directives.
+		slash := -1
+		for {
+			var rel string
+			i := strings.Index(relToVisit[slash+1:], "/")
+			if i < 0 {
+				rel = relToVisit
+			} else {
+				rel = relToVisit[:slash+1+i]
+				slash = slash + 1 + i
+			}
+
+			if _, ok := w.visits[rel]; !ok {
+				parentRel := path.Dir(rel)
+				if parentRel == "." {
+					parentRel = ""
+				}
+				parentCfg := w.visits[parentRel].c
+				if getWalkConfig(parentCfg).isExcludedDir(rel) {
+					break
+				}
+				if _, err := w.cache.get(rel, w.loadDirInfo); errors.Is(err, fs.ErrNotExist) {
+					// Directory does not exist.
+					break
+				}
+
+				c := parentCfg.Clone()
+				w.visit(c, rel, false)
+			}
+
+			if rel == relToVisit {
+				break
+			}
+		}
+
+		if c.Strict && len(w.errs) > 0 {
```

Comment:
Should this be checked after each `.visit` call? If not can we remove the check below the first `w.visit(c, "", false)` and just have this one (but moved out of the loop)?

Reply:
This makes us break out of the loop early in strict mode if there's an error. `walker.visit` does the same thing and doesn't recurse into subdirectories after an error. I think it's important to be consistent there.

(I'm not sure I really like strict mode, but Gazelle doesn't really provide a good framework for error handling in extensions, so perhaps it's needed.)

Reply:
Doesn't that mean it should be invoked directly after `w.visit()` and not down here then?

Reply:
Yes, follow-up in https://github.com/bazel-contrib/bazel-gazelle/pull/2077